source/extensions/filters/network/kafka/kafka_request.h

```diff
@@ -0,0 +1,337 @@
+    return os << "{header=" << arg.request_header_ << ", remaining=" << arg.remaining_request_size_
+              << "}";
+  }
+};
+
+typedef std::shared_ptr<RequestContext> RequestContextSharedPtr;
+
+/**
+ * Function generating a parser with given context
+ */
+typedef std::function<ParserSharedPtr(RequestContextSharedPtr)> ParserGeneratorFunction;
+
+/**
+ * Structure responsible for mapping [api_key, api_version] -> ParserGeneratorFunction
+ */
+typedef std::unordered_map<int16_t, std::unordered_map<int16_t, ParserGeneratorFunction>>
+    ParserGenerators;
+
+/**
+ * Trivial structure specifying which parser generator function should be used
+ * for which api_key & api_version
+ */
+struct ParserSpec {
+  const int16_t api_key_;
+  const std::vector<int16_t> api_versions_;
+  const ParserGeneratorFunction generator_;
+};
+
+/**
+ * Configuration object
+ * Resolves the parser that will be responsible for consuming the request-specific data
+ * In other words: provides (api_key, api_version) -> Parser function
+ */
+class RequestParserResolver {
+public:
+  /**
+   * Creates a resolver that uses generator functions provided by given specifications
+   */
+  RequestParserResolver(const std::vector<ParserSpec> specs);
+
+  /**
+   * Creates a resolver that uses generator functions provided by original resolver and then
+   * expanded by specifications
+   */
+  RequestParserResolver(const RequestParserResolver& original, const std::vector<ParserSpec> specs);
+
+  virtual ~RequestParserResolver() = default;
+
+  /**
+   * Creates a parser that is going to process data specific for given api_key & api_version
+   * @param api_key request type
+   * @param api_version request version
+   * @param context context to be used by parser
+   * @return parser that is capable of processing data for given request type & version
+   */
+  virtual ParserSharedPtr createParser(int16_t api_key, int16_t api_version,
+                                       RequestContextSharedPtr context) const;
+
+  /**
+   * Request versions handled by Kafka up to 0.11
+   */
+  static const RequestParserResolver KAFKA_0_11;
+
+  /**
+   * Request versions handled by Kafka up to 1.0
+   */
+  static const RequestParserResolver KAFKA_1_0;
+
+private:
+  ParserGenerators generators_;
+};
+
+/**
+ * Request parser responsible for consuming request length and setting up context with this data
+ * @see http://kafka.apache.org/protocol.html#protocol_common
+ */
+class RequestStartParser : public Parser {
+public:
+  RequestStartParser(const RequestParserResolver& parser_resolver)
+      : parser_resolver_{parser_resolver}, context_{std::make_shared<RequestContext>()} {};
+
+  /**
+   * Consumes INT32 bytes as request length and updates the context with that value
+   * @return RequestHeaderParser instance to process request header
+   */
+  ParseResponse parse(const char*& buffer, uint64_t& remaining);
```

Comment:
For overridden functions:

1) Use the `override` keyword
2) Don't duplicate the doc comment from the interface header
3) Precede the overrides of the comments with a comment similar to `// Extensions::Networkfilter::Kafka::Parser`

Please audit for this elsewhere.